File: python/advanced_ds/b_plus_tree.py
"""
B+ Tree (Database Index Structure)

Time Complexity:
- Search: O(log n)
- Insert: O(log n)
- Delete: O(log n)
- Range scan: O(log n + k) where k = results

Space Complexity: O(n)

Use Cases:
- Database indexing (primary index)
- File system indexing
- Multi-level sorted data structure
- Disk-based searching (minimizes I/O)
- Range queries over large datasets

Key Insight:
- All keys in leaves, internal nodes just guide search
- All leaves at same depth (balanced)
- Each node contains 60% to 100% of max capacity (t-1 to 2t-1 keys)
- Facilitates efficient range scans
- Leaf nodes can be linked for sequential access
"""

from typing import List, Optional, Tuple


class BPlusNode:
    """Node in B+ tree."""

    def __init__(self, is_leaf: bool, t: int):
        self.is_leaf = is_leaf
        self.t = t  # Minimum degree
        self.keys = []
        self.children = []
        self.next = None  # For leaf nodes: link to next leaf

    def is_full(self) -> bool:
        """Check if node is full."""
        return len(self.keys) == 2 * self.t - 1

    def has_min_keys(self) -> bool:
        """Check if node has minimum keys."""
        return len(self.keys) >= self.t - 1


class BPlusTree:
    """B+ tree for efficient searching and range queries."""

    def __init__(self, t: int = 3):
        """
        Initialize B+ tree.

        Args:
            t: Minimum degree (internal nodes have t-1 to 2t-1 keys)
        """
        self.root = BPlusNode(is_leaf=True, t=t)
        self.t = t

    def search(self, key: int) -> bool:
        """Search for key in tree."""
        return self._search(self.root, key) is not None

    def _search(self, node: BPlusNode, key: int) -> Optional[BPlusNode]:
        """Search and return leaf node containing key."""
        i = 0
        while i < len(node.keys) and key > node.keys[i]:
            i += 1

        if node.is_leaf:
            return node if i < len(node.keys) and node.keys[i] == key else None

        if i < len(node.keys) and key == node.keys[i]:
            return self._search(node.children[i + 1], key)

        return self._search(node.children[i], key)

    def insert(self, key: int) -> None:
        """Insert key into tree."""
        if self.root.is_full():
            new_root = BPlusNode(is_leaf=False, t=self.t)
            new_root.children.append(self.root)
            self._split_child(new_root, 0)
            self.root = new_root

        self._insert_non_full(self.root, key)

    def _insert_non_full(self, node: BPlusNode, key: int) -> None:
        """Insert into non-full node."""
        i = len(node.keys) - 1

        if node.is_leaf:
            # Find position and insert
            node.keys.append(None)
            while i >= 0 and key < node.keys[i]:
                node.keys[i + 1] = node.keys[i]
                i -= 1
            node.keys[i + 1] = key
        else:
            # Find child to recurse
            while i >= 0 and key < node.keys[i]:
                i -= 1
            i += 1

            if node.children[i].is_full():
                self._split_child(node, i)
                if key > node.keys[i]:
                    i += 1

            self._insert_non_full(node.children[i], key)

    def _split_child(self, parent: BPlusNode, idx: int) -> None:
        """Split child at index idx."""
        full_child = parent.children[idx]
        new_child = BPlusNode(is_leaf=full_child.is_leaf, t=self.t)

        mid = self.t - 1

        # Copy keys to new child
        new_child.keys = full_child.keys[mid + 1:]
        full_child.keys = full_child.keys[:mid]

        # If not leaf, copy children
        if not full_child.is_leaf:
            new_child.children = full_child.children[mid + 1:]
            full_child.children = full_child.children[:mid + 1]
        else:
            # Link leaf nodes
            new_child.next = full_child.next
            full_child.next = new_child

        # Move median key to parent
        parent.keys.insert(idx, full_child.keys[mid] if full_child.keys else new_child.keys[0])
        parent.children.insert(idx + 1, new_child)

    def range_search(self, lower: int, upper: int) -> List[int]:
        """Find all keys in range [lower, upper]."""
        result = []
        leaf = self._find_leaf(self.root, lower)

        while leaf:
            for key in leaf.keys:
                if lower <= key <= upper:
                    result.append(key)
                elif key > upper:
                    return result

            leaf = leaf.next

        return result

    def _find_leaf(self, node: BPlusNode, key: int) -> Optional[BPlusNode]:
        """Find leaf that could contain key."""
        i = 0
        while i < len(node.keys) and key > node.keys[i]:
            i += 1

        if node.is_leaf:
            return node

        return self._find_leaf(node.children[i], key)

    def delete(self, key: int) -> None:
        """Delete key from tree."""
        self._delete(self.root, key)

        if len(self.root.keys) == 0:
            if not self.root.is_leaf and len(self.root.children) > 0:
                self.root = self.root.children[0]

    def _delete(self, node: BPlusNode, key: int) -> None:
        """Delete key from node (simplified)."""
        i = 0
        while i < len(node.keys) and key > node.keys[i]:
            i += 1

        if node.is_leaf:
            if i < len(node.keys) and node.keys[i] == key:
                node.keys.pop(i)
        else:
            if i < len(node.keys) and node.keys[i] == key:
                # Merge children
                if len(node.children[i].keys) >= self.t:
                    node.keys[i] = node.children[i].keys[-1]
                    self._delete(node.children[i], node.keys[i])
                else:
                    self._delete(node.children[i + 1], key)
            else:
                self._delete(node.children[i], key)

    def inorder(self) -> List[int]:
        """Get all keys in order."""
        result = []
        self._inorder(self.root, result)
        return result

    def _inorder(self, node: BPlusNode, result: List[int]) -> None:
        """Inorder traversal."""
        if node.is_leaf:
            result.extend(node.keys)
        else:
            for i in range(len(node.keys)):
                self._inorder(node.children[i], result)
                result.append(node.keys[i])
            self._inorder(node.children[-1], result)


if __name__ == "__main__":
    # Example 1: Basic operations
    print("=== Example 1: Basic B+ Tree Operations ===")
    tree = BPlusTree(t=3)

    keys = [10, 20, 5, 6, 12, 30, 7, 17]
    print(f"Inserting: {keys}")
    for key in keys:
        tree.insert(key)

    print(f"Inorder: {tree.inorder()}")

    for key in keys + [8, 100]:
        print(f"Search {key}: {tree.search(key)}")

    # Example 2: Range search
    print("\n=== Example 2: Range Search ===")
    print(f"Range [5, 20]: {tree.range_search(5, 20)}")
    print(f"Range [7, 15]: {tree.range_search(7, 15)}")

    # Example 3: Larger tree
    print("\n=== Example 3: Larger Tree ===")
    tree2 = BPlusTree(t=3)
    for i in range(1, 21):
        tree2.insert(i)

    print(f"Inorder (1-20): {tree2.inorder()}")
    print(f"Range [5, 15]: {tree2.range_search(5, 15)}")

File: java/advanced_ds/BPlusTree.java
package advanced_ds;

import java.util.*;

/**
 * B+ Tree (Database Index Structure)
 *
 * Time Complexity:
 * - Search: O(log n)
 * - Insert: O(log n)
 * - Delete: O(log n)
 * - Range scan: O(log n + k) where k = results
 *
 * Space Complexity: O(n)
 *
 * Use Cases:
 * - Database indexing (primary index)
 * - File system indexing
 * - Multi-level sorted data structure
 * - Disk-based searching (minimizes I/O)
 * - Range queries over large datasets
 *
 * Key Insight:
 * - All keys in leaves, internal nodes just guide search
 * - All leaves at same depth (balanced)
 * - Each node contains 60% to 100% of max capacity (t-1 to 2t-1 keys)
 * - Facilitates efficient range scans
 * - Leaf nodes can be linked for sequential access
 */
public class BPlusTree {

    static class Node {
        boolean isLeaf;
        List<Integer> keys;
        List<Node> children;
        Node next;  // For leaf nodes
        int t;      // Minimum degree

        Node(boolean isLeaf, int t) {
            this.isLeaf = isLeaf;
            this.t = t;
            this.keys = new ArrayList<>();
            this.children = new ArrayList<>();
        }

        boolean isFull() {
            return keys.size() == 2 * t - 1;
        }

        boolean hasMinKeys() {
            return keys.size() >= t - 1;
        }
    }

    private Node root;
    private int t;

    /**
     * Initialize B+ tree.
     *
     * @param t Minimum degree
     */
    public BPlusTree(int t) {
        this.root = new Node(true, t);
        this.t = t;
    }

    /**
     * Search for key in tree.
     *
     * @param key Key to search
     * @return true if key exists
     */
    /**
     * [Brief description]
     *
     * @param [param] [description]
     * @return [description]
     * @time O([complexity])
     */
    public boolean search(int key) {
        return search(root, key) != null;
    }

    /**
     * Search and return leaf node containing key.
     */
    private Node search(Node node, int key) {
        int i = 0;
        while (i < node.keys.size() && key > node.keys.get(i)) {
            i++;
        }

        if (node.isLeaf) {
            return (i < node.keys.size() && node.keys.get(i) == key) ? node : null;
        }

        if (i < node.keys.size() && key == node.keys.get(i)) {
            return search(node.children.get(i + 1), key);
        }

        return search(node.children.get(i), key);
    }

    /**
     * Insert key into tree.
     *
     * @param key Key to insert
     */
    public void insert(int key) {
        if (root.isFull()) {
            Node newRoot = new Node(false, t);
            newRoot.children.add(root);
            splitChild(newRoot, 0);
            root = newRoot;
        }

        insertNonFull(root, key);
    }

    private void insertNonFull(Node node, int key) {
        int i = node.keys.size() - 1;

        if (node.isLeaf) {
            node.keys.add(null);
            while (i >= 0 && key < node.keys.get(i)) {
                node.keys.set(i + 1, node.keys.get(i));
                i--;
            }
            node.keys.set(i + 1, key);
        } else {
            while (i >= 0 && key < node.keys.get(i)) {
                i--;
            }
            i++;

            if (node.children.get(i).isFull()) {
                splitChild(node, i);
                if (key > node.keys.get(i)) {
                    i++;
                }
            }

            insertNonFull(node.children.get(i), key);
        }
    }

    private void splitChild(Node parent, int idx) {
        Node fullChild = parent.children.get(idx);
        Node newChild = new Node(fullChild.isLeaf, t);

        int mid = t - 1;

        // Copy keys
        for (int i = mid + 1; i < fullChild.keys.size(); i++) {
            newChild.keys.add(fullChild.keys.get(i));
        }
        fullChild.keys = new ArrayList<>(fullChild.keys.subList(0, mid));

        // Copy children if not leaf
        if (!fullChild.isLeaf) {
            for (int i = mid + 1; i < fullChild.children.size(); i++) {
                newChild.children.add(fullChild.children.get(i));
            }
            fullChild.children = new ArrayList<>(fullChild.children.subList(0, mid + 1));
        } else {
            // Link leaf nodes
            newChild.next = fullChild.next;
            fullChild.next = newChild;
        }

        // Move median to parent
        int medianKey = newChild.keys.isEmpty() ?
                (fullChild.keys.isEmpty() ? -1 : fullChild.keys.get(fullChild.keys.size() - 1)) :
                newChild.keys.get(0);

        parent.keys.add(idx, medianKey);
        parent.children.add(idx + 1, newChild);
    }

    /**
     * Find all keys in range [lower, upper].
     *
     * @param lower Lower bound
     * @param upper Upper bound
     * @return Keys in range
     */
    /**
     * [Brief description]
     *
     * @param [param] [description]
     * @return [description]
     * @time O([complexity])
     */
    public List<Integer> rangeSearch(int lower, int upper) {
        List<Integer> result = new ArrayList<>();
        Node leaf = findLeaf(root, lower);

        while (leaf != null) {
            for (int key : leaf.keys) {
                if (lower <= key && key <= upper) {
                    result.add(key);
                } else if (key > upper) {
                    return result;
                }
            }
            leaf = leaf.next;
        }

        return result;
    }

    private Node findLeaf(Node node, int key) {
        int i = 0;
        while (i < node.keys.size() && key > node.keys.get(i)) {
            i++;
        }

        if (node.isLeaf) {
            return node;
        }

        return findLeaf(node.children.get(i), key);
    }

    /**
     * Get all keys in inorder.
     *
     * @return Sorted list of keys
     */
    public List<Integer> inorder() {
        List<Integer> result = new ArrayList<>();
        inorder(root, result);
        return result;
    }

    private void inorder(Node node, List<Integer> result) {
        if (node.isLeaf) {
            result.addAll(node.keys);
        } else {
            for (int i = 0; i < node.keys.size(); i++) {
                inorder(node.children.get(i), result);
                result.add(node.keys.get(i));
            }
            inorder(node.children.get(node.children.size() - 1), result);
        }
    }

    /**
     * [Brief description]
     *
     * @param [param] [description]
     * @return [description]
     * @time O([complexity])
     */
    public static void main(String[] args) {
        // Example 1: Basic operations
        System.out.println("=== Example 1: Basic B+ Tree Operations ===");
        BPlusTree tree = new BPlusTree(3);

        int[] keys = {10, 20, 5, 6, 12, 30, 7, 17};
        System.out.print("Inserting: [");
        for (int i = 0; i < keys.length; i++) {
            if (i > 0) System.out.print(", ");
            System.out.print(keys[i]);
            tree.insert(keys[i]);
        }
        System.out.println("]");

        System.out.println("Inorder: " + tree.inorder());

        for (int key : new int[]{5, 10, 12, 8, 100}) {
            System.out.println("Search " + key + ": " + tree.search(key));
        }

        // Example 2: Range search
        System.out.println("\n=== Example 2: Range Search ===");
        System.out.println("Range [5, 20]: " + tree.rangeSearch(5, 20));
        System.out.println("Range [7, 15]: " + tree.rangeSearch(7, 15));

        // Example 3: Larger tree
        System.out.println("\n=== Example 3: Larger Tree ===");
        BPlusTree tree2 = new BPlusTree(3);
        for (int i = 1; i <= 20; i++) {
            tree2.insert(i);
        }

        System.out.println("Inorder (1-20): " + tree2.inorder());
        System.out.println("Range [5, 15]: " + tree2.rangeSearch(5, 15));
    }
}
